File: Species/Species.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <set>
#include "Species.h"
#include "Type.h"
#include "json.hpp"
using json = nlohmann::json;

Species::Species(const json& inputJSON) {
    _name = inputJSON.at("name").get<std::string>();

    _stats["HitPoints"] = inputJSON.at("stats").at("HitPoints").get<int>();
    _stats["Attack"] = inputJSON.at("stats").at("Attack").get<int>();
    _stats["Defense"] = inputJSON.at("stats").at("Defense").get<int>();
    _stats["SpecialAttack"] = inputJSON.at("stats").at("SpecialAttack").get<int>();
    _stats["SpecialDefense"] = inputJSON.at("stats").at("SpecialDefense").get<int>();
    _stats["Speed"] = inputJSON.at("stats").at("Speed").get<int>();

    _effortValues["HitPoints"] = inputJSON.at("effort-values").at("HitPoints").get<int>();
    _effortValues["Attack"] = inputJSON.at("effort-values").at("Attack").get<int>();
    _effortValues["Defense"] = inputJSON.at("effort-values").at("Defense").get<int>();
    _effortValues["SpecialAttack"] = inputJSON.at("effort-values").at("SpecialAttack").get<int>();
    _effortValues["SpecialDefense"] = inputJSON.at("effort-values").at("SpecialDefense").get<int>();
    _effortValues["Speed"] = inputJSON.at("effort-values").at("Speed").get<int>();

    _type1 = &typeMap.find(inputJSON.at("types").at("1").get<std::string>())->second;
    _type2 = &typeMap.find(inputJSON.at("types").at("2").get<std::string>())->second;

    _abilities[1] = inputJSON.at("abilities").at("1").get<std::string>();
    _abilities[2] = inputJSON.at("abilities").at("2").get<std::string>();
    _abilities[3] = inputJSON.at("abilities").at("3").get<std::string>();

    _evolution = inputJSON.at("evolve").at("evolves-to").get<std::string>();

    _growthRate = inputJSON.at("growth-rate").get<std::string>();

    _experienceValue = inputJSON.at("base-experience").get<int>();

    _moves.levelUp = inputJSON.at("moves").at("level-up");
    _moves.egg = inputJSON.at("moves").at("egg");
    _moves.machine = inputJSON.at("moves").at("machine");
    _moves.tutor = inputJSON.at("moves").at("tutor");
    _moves.other = inputJSON.at("moves").at("other");
}

std::map<std::string, Species> speciesMap;

void generateSpeciesMap() {
    std::string filePath = "Species.json";
    std::ifstream ifs(filePath);
    if (!ifs.is_open()) {
        std::cerr << "Error opening file: " << filePath << std::endl;
    }
    json speciesFile;
    ifs >> speciesFile;

    for (json::iterator specie = speciesFile.begin(); specie != speciesFile.end(); ++specie) {
        std::string speciesName = specie.key();
        json speciesJSON = specie.value();

        speciesMap.emplace(speciesName, Species(speciesJSON));
    }
}

File: Pokemon Black2 Replica/Assets/Pokemon/Species.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.CompilerServices;
using UnityEngine;

public class Species
{
    // Automatic Properties
    [JsonProperty("name")]
    public string Name { get; private set; } = string.Empty;

    [JsonProperty("stats")]
    public Dictionary<string, int> Stats { get; } = new Dictionary<string, int>();

    [JsonProperty("effort-values")]
    public Dictionary<string, int> EffortValues { get; } = new Dictionary<string, int>();

    public Type Type1 { get; private set; }
    public Type Type2 { get; private set; }
    [JsonProperty("types")]
    private Dictionary<int, string> TypeSetter
    {
        set
        {
            Type1 = Type.TypeMap[value[1]];
            Type2 = Type.TypeMap[value[2]];
        }
    }

    [JsonProperty("abilities")]
    public Dictionary<int, string> Abilities { get; } = new Dictionary<int, string>();

    public Species Evolution { get; private set; }
    private string EvolutionName { get; set; } = string.Empty;
    public int EvolutionLevel { get; private set; } = 0;
    [JsonProperty("evolve")]
    private Dictionary<string, object> EvolveSetter
    {
        set
        {
            EvolutionName = value["evolves-to"].ToString();
            EvolutionLevel = Convert.ToInt32(value["min-level"]);
        }
    }

    public GrowthRate GrowthRate { get; private set; }
    [JsonProperty("growth-rate")]
    private string TempGrowthRate { set { GrowthRate = GrowthRate.GrowthRateMap[value]; } }

    [JsonProperty("base-experience")]
    public int ExperienceValue { get; private set; } = 0;

    [JsonProperty("moves")]
    public LearnSet Moves { get; private set; }


    // Class for storing the Moves data
    public class LearnSet
    {
        public Dictionary<Move, int> LevelUp { get; private set; }
        [JsonProperty("level-up")]
        private Dictionary<string, int> LevelUpSetter { set { foreach (var (moveName, moveLevel) in value) { LevelUp[Move.MoveMap[moveName]] = moveLevel; } } }

        public HashSet<Move> Egg {  get; private set; }
        [JsonProperty("egg")]
        private HashSet<string> EggSetter { set { foreach (string moveName in value) Egg.Add(Move.MoveMap[moveName]); } }

        public HashSet<Move> Machine { get; private set; }
        [JsonProperty("machine")]
        private HashSet<string> MachineSetter { set { foreach (string moveName in value) Machine.Add(Move.MoveMap[moveName]); } }

        public HashSet<Move> Tutor { get; private set; }
        [JsonProperty("tutor")]
        private HashSet<string> TutorSetter { set { foreach (string moveName in value) Tutor.Add(Move.MoveMap[moveName]);  } }

        public Dictionary<Move, string> Other { get; private set; }
        [JsonProperty("other")]
        private Dictionary<string, string> OtherSetter { set { foreach (var (moveName, learnCondition) in value) { Other[Move.MoveMap[moveName]] = learnCondition; } } }

        private LearnSet()
        {
            LevelUp = new Dictionary<Move, int>();
            Egg = new HashSet<Move>();
            Machine = new HashSet<Move>();
            Tutor = new HashSet<Move>();
            Other = new Dictionary<Move, string>();
        }
    }

    // Establish Species
    public static Dictionary<string, Species> SpeciesMap { get; } = new Dictionary<string, Species>();

    static Species()
    {
        string filePath = Path.Combine(Application.streamingAssetsPath, "Data", "Species.json");
        string jsonString = File.ReadAllText(filePath);

        SpeciesMap = JsonConvert.DeserializeObject<Dictionary<string, Species>>(jsonString);

        foreach (Species species in SpeciesMap.Values)
        {
            species.Evolution = (species.EvolutionName == "Special") ? species : SpeciesMap[species.EvolutionName];
        }
    }
}
